### PR TITLE
TDL-6148: Catch and retry Attribute error and TDL-13267: Fix AdCreative requests that are not retrying

### PR DESCRIPTION
**Pull request by savan-chovatiya**
# Description of change
[TDL-6148](https://jira.talendforge.org/browse/TDL-6148): catch and retry 'AttributeError: 'str' object has no attribute 'get''
- Added AttributeError in retry_pattern  decorator for all the sync related calls.
- Also, added AttributeError into  should_retry_api_error for catching.

[TDL-13267](https://jira.talendforge.org/browse/TDL-13267): AdCreative requests that are not retrying
- Added retry_pattern  decorator on the sync_batches() function of AdCreatives with FacebookRequestError also.

# Manual QA steps
 - Added unit test case for verifying retries on the above exceptions
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_facebook/__init__.py [hunk 1]
```diff
@@ -153,6 +153,8 @@ def should_retry_api_error(exception):
             return True
         elif isinstance(exception, TypeError) and str(exception) == "string indices must be integers":
             return True
+        elif isinstance(exception, AttributeError):
```
Comment (dbshah1212) [inline tap_facebook/__init__.py:156]: We can add it in the or of `if isinstance(exception, FacebookBadObjectError)`, instead of adding a new elif.

Comment (savan-chovatiya) [reply tap_facebook/__init__.py]: Updates as per suggestion.

File: tests/unittests/test_sync_batches_retry.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
+import unittest
```
Comment (karanpanchal-crest) [inline tests/unittests/test_sync_batches_retry.py:1]: @savan-chovatiya What is the difference between the test cases written in this file and `test_attribute_error_retry.py` file.
In the above file as well, you are checking that call_count should be 5 times. What else you are checking in `test_sync_batches_retry` file?

File: tests/unittests/test_sync_batches_retry.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        # Initialize AdCreative and mock catalog_entry
+        mock_catalog_entry = CatalogEntry(schema=Schema())
+        ad_creative_object = AdCreative('', '', '', '')
+        ad_creative_object.catalog_entry = mock_catalog_entry
+
+        # Call sync_batches() function of AdCreatives and verify AttributeError is raised
+        with self.assertRaises(AttributeError):
+            ad_creative_object.sync_batches([])
+
+        # verify calls inside sync_batches are called 5 times as max 5 reties provided for function
```
Comment (karanpanchal-crest) [inline tests/unittests/test_sync_batches_retry.py:51]: Typo: It should be 'retries'

**Review by karanpanchal-crest — CHANGES_REQUESTED** [review]

**Comment by karanpanchal-crest** [comment]
@savan-chovatiya Please add comments to which code change corresponds to which JIRA

File: tests/unittests/test_sync_batches_retry.py (hunk 2 above)
Comment (savan-chovatiya) [reply tests/unittests/test_sync_batches_retry.py:1]: There was no backoff on the `sync_batches` function so I added and also include the `FacebookRequestError` error with `AttributeError` error so those tests are under `test_sync_batches_retry`. While other test cases for `AttributeError` are under 'test_attribute_error_retry.py'

File: tests/unittests/test_sync_batches_retry.py (hunk 3 above)
Comment (savan-chovatiya) [reply tests/unittests/test_sync_batches_retry.py]: Fixed the typo.

File: tap_facebook/__init__.py [hunk 4]
```diff
@@ -241,6 +241,7 @@ class AdCreative(Stream):
     doc: https://developers.facebook.com/docs/marketing-api/reference/adgroup/adcreatives/
     '''
 
+    @retry_pattern(backoff.expo, (FacebookRequestError, AttributeError), max_tries=5, factor=5)
```
Comment (savan-chovatiya) [inline tap_facebook/__init__.py:244]: This change is for [TDL-13267](https://jira.talendforge.org/browse/TDL-13267): AdCreative requests that are not retrying on `AttributeError`

**Review by karanpanchal-crest — APPROVED** [review]

**Review by KrisPersonal — CHANGES_REQUESTED** [review]
These test cases needs to be reviewed by Kyle speer. Please assign this PR to him

**Review by KrisPersonal — CHANGES_REQUESTED** [review]
Add Change comments in the code

**Comment by savan-chovatiya** [comment]
> Add Change comments in the code

Added comments for changes in code.

File: tests/unittests/test_attribute_error_retry.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+        # Call get_adcreatives() function of AdCreatives and verify AttributeError is raised
+        ad_creative_object = AdCreative('', mocked_account, '', '')
+        with self.assertRaises(AttributeError):
+            ad_creative_object.get_adcreatives()
+
+        # verify get_ad_creatives() is called 5 times as max 5 reties provided for function
+        self.assertEquals(mocked_account.get_ad_creatives.call_count, 5)
+
+    def test__call_get_ads(self, mocked_sleep):
```
Comment (kspeer825) [inline tests/unittests/test_attribute_error_retry.py:29]:
```suggestion
    def test_call_get_ads(self, mocked_sleep):
```

**Review by kspeer825 — APPROVED** [review]
Left a suggestion

File: tests/unittests/test_attribute_error_retry.py (hunk 5 above)
Comment (savan-chovatiya) [reply tests/unittests/test_attribute_error_retry.py]: Updated test name as per suggestion.

**Review by KrisPersonal — APPROVED** [review]

**Pull request merged**